[OOB 5.6 r2] make_armc5 exporter does not support NUCLEO_F767ZI, NUCLEO_F401RE and DISCO_L476VG

### Description
- Type: Bug
- Priority: Minor

---------------------------------------------------------------
## Bug

**Target**
NUCLEO_F767ZI, NUCLEO_F401RE and DISCO_L476VG|??

**Toolchain:**
ARMC5

**Toolchain version:**
MDK_ARM Professional Ver. : 5.22
C Compiler: v5.06 update 4(build 422)

**mbed-cli version:** 
1.2.0

**mbed-os sha:**
mbed-os-example-fat-filesystem (08607866ff3f09b75e429ac0432587478d5e1fc3)
mbed-os (6e087488e10143222b3b08790eb45600c784a309)

**Expected behavior**
Make can pass to generate a binary executable without error.

**Actual behavior**
Make get failed. please check the attached log.

> "Assemble: irq_cm4f.s"
> "../mbed-os/rtos/TARGET_CORTEX/rtx5/TARGET_RTOS_M4_M7/TOOLCHAIN_ARM/irq_cm4f.s", line 69 (column 8): Error: A1163E: Unknown opcode __FPU_PRESENT , expecting opcode or Macro
>    69 00000030 #ifdef __FPU_PRESENT
>                       ^
> "../mbed-os/rtos/TARGET_CORTEX/rtx5/TARGET_RTOS_M4_M7/TOOLCHAIN_ARM/irq_cm4f.s", line 75: Error: A1167E: Invalid line start 
>    75 0000003c #endif
> "../mbed-os/rtos/TARGET_CORTEX/rtx5/TARGET_RTOS_M4_M7/TOOLCHAIN_ARM/irq_cm4f.s", line 79 (column 8): Error: A1163E: Unknown opcode __FPU_PRESENT , expecting opcode or Macro
>    79 00000040 #ifdef __FPU_PRESENT
>                       ^
> "../mbed-os/rtos/TARGET_CORTEX/rtx5/TARGET_RTOS_M4_M7/TOOLCHAIN_ARM/irq_cm4f.s", line 82: Error: A1167E: Invalid line start 
>    82 0000004a #endif
> "../mbed-os/rtos/TARGET_CORTEX/rtx5/TARGET_RTOS_M4_M7/TOOLCHAIN_ARM/irq_cm4f.s", line 95 (column 8): Error: A1163E: Unknown opcode __FPU_PRESENT , expecting opcode or Macro
>    95 0000005e #ifdef __FPU_PRESENT
>                       ^
> "../mbed-os/rtos/TARGET_CORTEX/rtx5/TARGET_RTOS_M4_M7/TOOLCHAIN_ARM/irq_cm4f.s", line 98: Error: A1167E: Invalid line start 
>    98 00000068 #endif
> 6 Errors, 0 Warnings
> make[1]: *** [C:/ioT/testpod/oob5.6/mbed-os-example-fat-filesystem/Makefile:779: mbed-os/rtos/TARGET_CORTEX/rtx5/TARGET_RTOS_M4_M7/TOOLCHAIN_ARM/irq_cm4f.o] Error 1
> make: *** [makefile:26: all] Error 2

**Steps to reproduce**

> \>mbed import mbed-os-example-fat-filesystem
> \>cd mbed-os-example-fat-filesystem
> \>mbed update mbed-os-5.6-oob2
> \>mbed export -m NUCLEO_F767ZI -i make_armc5 -vv
> \>make
[failed-make-F767ZI.txt](https://github.com/ARMmbed/mbed-os/files/1317102/failed-make-F767ZI.txt)

--- a/tools/export/makefile/Makefile.tmpl
+++ b/tools/export/makefile/Makefile.tmpl
@@ -97,20 +97,26 @@ all: $(PROJECT).bin $(PROJECT)-combined.hex size
 all: $(PROJECT).bin $(PROJECT).hex size
 {% endif %}
 
-.asm.o:
-	+@$(call MAKEDIR,$(dir $@))
-	+@echo "Assemble: $(notdir $<)"
-	@$(AS) -c $(ASM_FLAGS) $(INCLUDE_PATHS) -o $@ $<
-
 .s.o:
 	+@$(call MAKEDIR,$(dir $@))
 	+@echo "Assemble: $(notdir $<)"
+  {% if needs_asm_preproc %}
+	@$(AS) -c $(ASM_FLAGS) $(INCLUDE_PATHS) -E -o $(@:.o=.E.s) $<
+	@$(AS) -c $(ASM_FLAGS) $(INCLUDE_PATHS) -o $@ $(@:.o=.E.s)
+  {% else %}
 	@$(AS) -c $(ASM_FLAGS) $(INCLUDE_PATHS) -o $@ $<
+  {% endif %}
+
 
 .S.o:
 	+@$(call MAKEDIR,$(dir $@))
 	+@echo "Assemble: $(notdir $<)"
+  {% if needs_asm_preproc %}
+	@$(AS) -c $(ASM_FLAGS) $(INCLUDE_PATHS) -E -o $(@:.o=.E.s) $<
+	@$(AS) -c $(ASM_FLAGS) $(INCLUDE_PATHS) -o $@ $(@:.o=.E.s)
+  {% else %}
 	@$(AS) -c $(ASM_FLAGS) $(INCLUDE_PATHS) -o $@ $<
+  {% endif %}
 
 .c.o:
 	+@$(call MAKEDIR,$(dir $@))

--- a/tools/export/makefile/__init__.py
+++ b/tools/export/makefile/__init__.py
@@ -35,6 +35,8 @@ class Makefile(Exporter):
 
     MBED_CONFIG_HEADER_SUPPORTED = True
 
+    PREPROCESS_ASM = False
+
     POST_BINARY_WHITELIST = set([
         "MCU_NRF51Code.binary_hook",
         "TEENSY3_1Code.binary_hook",
@@ -96,6 +98,7 @@ class Makefile(Exporter):
             'link_script_ext': self.toolchain.LINKER_EXT,
             'link_script_option': self.LINK_SCRIPT_OPTION,
             'user_library_flag': self.USER_LIBRARY_FLAG,
+            'needs_asm_preproc': self.PREPROCESS_ASM,
         }
 
         if hasattr(self.toolchain, "preproc"):
@@ -236,6 +239,7 @@ class Armc5(Arm):
     """ARM Compiler 5 (armcc) specific makefile target"""
     NAME = 'Make-ARMc5'
     TOOLCHAIN = "ARM"
+    PREPROCESS_ASM = True
 
 class Armc6(Arm):
     """ARM Compiler 6 (armclang) specific generic makefile target"""

